File: MONATE.Web.Server/Helpers/CryptionHelper.cs
namespace MONATE.Web.Server.Helpers
{
    using System;
    using System.Text;
    using Sodium;

    public class CryptionHelper
    {
        private readonly byte[] key;

        public CryptionHelper(string passwordHex)
        {
            // Convert hex string to byte array
            this.key = Convert.FromHexString(passwordHex);
        }

        public string Encrypt(string plaintext)
        {
            // Generate random nonce (IV) and associated data
            byte[] nonce = SodiumCore.GetRandomBytes(12); // 12 bytes
            byte[] associatedData = SodiumCore.GetRandomBytes(16); // 16 bytes

            // Encrypt the plaintext
            byte[] cipherText = SecretAeadChaCha20Poly1305.Encrypt(
                Encoding.UTF8.GetBytes(plaintext),
                associatedData,
                nonce,
                key
            );

            // Convert components to hex for easy transport
            string nonceHex = Convert.ToHexString(nonce).ToLower();
            string associatedDataHex = Convert.ToHexString(associatedData).ToLower();
            string cipherTextHex = Convert.ToHexString(cipherText).ToLower();

            return nonceHex + associatedDataHex + cipherTextHex;
        }

        public string Decrypt(string cipherTextHex)
        {
            // Extract the nonce, associated data, and cipher text from the hex string
            byte[] nonce = Convert.FromHexString(cipherTextHex[0..24]); // First 12 bytes
            byte[] associatedData = Convert.FromHexString(cipherTextHex[24..32]); // Next 16 bytes
            byte[] encryptedData = Convert.FromHexString(cipherTextHex[56..]); // Remaining bytes

            // Decrypt the cipher text
            byte[] decrypted = SecretAeadChaCha20Poly1305.Decrypt(
                encryptedData,
                associatedData,
                nonce,
                key
            );

            return Encoding.UTF8.GetString(decrypted);
        }
    }
}

File: monate.web.client/helpers/cryption-helper.js
import sodium from 'libsodium-wrappers-sumo';

class CryptionHelper {
    encoding = 'hex';
    key = null;

    async initialize() {
        await sodium.ready; // Ensure libsodium is ready
        const passwordHex = sessionStorage.getItem('password');
        if (!passwordHex) {
            throw new Error("Password not found in session storage.");
        }
        this.key = sodium.from_hex(passwordHex); // Convert hex string to Uint8Array
    }

    async encrypt(plaintext) {
        if (!this.key) {
            throw new Error("Key not initialized. Ensure libsodium is ready and key is set.");
        }
        try {
            const iv = sodium.randombytes_buf(12); // Random 12-byte nonce
            const assocData = sodium.randombytes_buf(16); // Random associated data

            const cipherText = sodium.crypto_aead_chacha20poly1305_ietf_encrypt(
                sodium.from_string(plaintext), // Message to encrypt
                assocData,                     // Associated data (optional)
                null,                          // No additional data used
                iv,                            // Nonce (IV)
                this.key                       // Secret key
            );

            return sodium.to_hex(iv) + sodium.to_hex(assocData) + sodium.to_hex(cipherText);
        } catch (e) {
            console.error(e);
        }
    }

    async decrypt(cipherText) {
        if (!this.key) {
            throw new Error("Key not initialized. Ensure libsodium is ready and key is set.");
        }
        try {
            const iv = sodium.from_hex(cipherText.slice(0, 24)); // First 12 bytes for nonce
            const assocData = sodium.from_hex(cipherText.slice(24, 56)); // Next 16 bytes for AAD
            const encryptedData = sodium.from_hex(cipherText.slice(56)); // Remaining bytes for encrypted data

            const decrypted = sodium.crypto_aead_chacha20poly1305_ietf_decrypt(
                null,                // No additional data
                encryptedData,      // Encrypted message
                assocData,          // Associated data (AAD)
                iv,                 // Nonce (IV)
                this.key            // Secret key
            );

            return sodium.to_string(decrypted); // Convert decrypted Uint8Array back to string
        } catch (e) {
            console.error(e);
        }
    }
}

export default CryptionHelper;
